### PR TITLE
DXF: Improved handling of very long lines of text in DXF file

### PR DESCRIPTION
**Pull request by atlight**
Although lines in a DXF file are supposed to be limited to (about) 256 columns, in practice they can be longer than this. Two-character escape sequences like ^I are counted as a single character, while some DXF group codes (such as MLEADER group code 304) exceed the limit because the format does not provide a generalised way to wrap long text strings.

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

**Comment by atlight** [comment]
This should fix #3828

**Comment by rouault** [comment]
(for a next time, could you push your branch to another remote than OSGeo/gdal when issuing the PR ? Otherwise all checks are done twice here: as a pull request and as a branch push)

**Comment by atlight** [comment]
> (for a next time, could you push your branch to another remote than OSGeo/gdal when issuing the PR ? Otherwise all checks are done twice here: as a pull request and as a branch push)

Yes, not a problem. Thanks for pointing that out, Even.

File: gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp [hunk 1]
```diff
@@ -179,13 +177,57 @@ int OGRDXFReader::ReadValueRaw( char *pszValueBuf, int nValueBufSize )
            && achSrcBuffer[iEOL] != '\0' )
         iEOL++;
 
-    if( achSrcBuffer[iEOL] == '\0' )
-        return -1;
+    while( achSrcBuffer[iEOL] == '\0' )
+    {
+        // The line is longer than the buffer. Let's copy what we have so
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp:182]: is this situation tested in the tests?

File: gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp [hunk 2]
```diff
@@ -179,13 +177,57 @@ int OGRDXFReader::ReadValueRaw( char *pszValueBuf, int nValueBufSize )
...
+        // far into our string, and read more
+        const auto nValueLength = osValue.length();
+        osValue.resize( nValueLength + iEOL - iSrcBufferOffset, '\0' );
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp:185]: couldn't this grow out of control (thinking of potential hostile files) ? Shoudn't we had a check up to ~ 1MB of osValue.size() ?

File: gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp [hunk 3]
```diff
@@ -179,13 +177,57 @@ int OGRDXFReader::ReadValueRaw( char *pszValueBuf, int nValueBufSize )
...
+        if( iEOL == iSrcBufferOffset )
+            break;
+    }
+
+    size_t nValueBufLen = 0;
+
+    // If this was an extremely long line, copy from osValue into the buffer
+    if( !osValue.empty() )
+    {
+        strncpy( pszValueBuf + nValueBufLen,
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp:210]:
```suggestion
        strncpy( pszValueBuf,
```

File: gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp [hunk 4]
```diff
@@ -179,13 +177,57 @@ int OGRDXFReader::ReadValueRaw( char *pszValueBuf, int nValueBufSize )
...
 
-    if( (iEOL - iSrcBufferOffset) > nValueBufSize-1 )
+    // Copy the last (normally, the only) section of this line into the buffer
+    if( (iEOL - iSrcBufferOffset) > nValueBufSize - (int)nValueBufLen - 1 )
     {
-        strncpy( pszValueBuf, achSrcBuffer + iSrcBufferOffset,
-                 nValueBufSize-1 );
+        strncpy( pszValueBuf + nValueBufLen,
+                 achSrcBuffer + iSrcBufferOffset,
+                 nValueBufSize - strlen(pszValueBuf) - 1 );
```
Comment (rouault) [inline gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp:230]:
```suggestion
                 nValueBufSize - static_cast<int>(nValueBufLen) - 1 );
```

File: gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp (hunk 2 above)
Comment (atlight) [reply gdal/ogr/ogrsf_frmts/dxf/ogrdxfreader.cpp:193]: The code now bails out if osValue is going to exceed 1MB.

**Comment by rouault** [comment]
@atlight https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=34927 has just been reported and it triggers in the OGRDXFReader::ReadValueRaw() method modified by this pull request

**Pull request merged**